_posts/pl/2021-01-11-rest-api-dobre-praktyki.md

```diff
@@ -0,0 +1,242 @@
+
+- Tracimy łatwe wsparcie dla cache'owania. Co prawda standard RFC 7234 zakłada wsparcie dla metod GET, HEAD i POST, ale większość implementacji wspiera tylko metody GET i HEAD.
+- Tracimy wsparcie dla ponawiania żądania przez kliencką bibliotekę HTTP przy błędzie sieciowym.
+- Tracimy zgodność ze standardem. Można też spierać się o to, jak bardzo to rozwiązanie jest nieeleganckie lub czy nie wprowadzi osób korzystających z naszego API w błąd.
+
+```
+POST /locations/search
+{
+    "country": "France,Italy,Greece,......,Brazil",
+    "minimum_population": 2500000,
+    "type": "city, village",
+    "max_height": 1300,
+    "min_area": 100000,
+    ....
+}
+```
+
+```
+"locations": [
+"id": "2443", "name": "Rome",
+"id": "6789", "name": "Paris",
+...
+]
+```
+
+Czy takie rozwiązanie są spotykane w praktyce? Tak, np. zdecydowali się na to twórcy Dropboxa, argumentując swoją decyzję tutaj.
+
+### GET i POST
+Trzeci pomysł to próba połączenia dwóch poprzednich rozwiązań w taki sposób, żebyśmy byli zgodni ze standardem i intuicyjnym rozumieniem metod (czyli do pobierania danych ma służyć GET, a nie POST). Najpierw tworzymy zapytanie wyszukiwania, przekazując w ciele POST-a parametry wyszukiwania. W odpowiedzi otrzymujemy id naszego wyszukiwania. Korzystamy z GETa i ze zwróconego wcześniej id, aby otrzymać nasz wynik.
+
+- Wprowadzamy nadmiernie skomplikowany mechanizm, a przecież chcemy jedynie otrzymać wynik wyszukiwania. W tym celu konieczne będą aż dwa wywołania usług i cały mechanizm na serwerze zapisujący gdzieś te wyniki wyszukiwania.
+- Można w tej koncepcji wprowadzić cache'owanie.
+- Jesteśmy zgodni ze standardem.
+
+```
+POST /locations/search
+{
+    "country": "France,Italy,Greece,......,Brazil",
+    "minimum_population": 2500000,
+    "type": "city, village",
+    "max_height": 1300,
+    "min_area": 100000,
+    ....
+}
+```
+
+```
+"id": "1297612456456"
+```
+
+```
+GET /locations/search?id=1297612456456
+```
+
+```
+"locations": [
+"id": "2443", "name": "Rome",
+"id": "6789", "name": "Paris",
+...
+]
+```
+
+Na pewno nie jest to popularne rozwiązanie, jest najbardziej kosztowne i ryzykowne, a zyskujemy zgodność ze standardem.
+
+
+## Obsługa błędów
+### Rozróżnianie między niewłaściwą ścieżką, nieistniejącym zasobem, brakiem uprawnień...
+
+Częstym błędem w odpowiedzi na żądania jest zwracanie nieprawidłowego (niezgodnego ze standardem) kodu błędu.  Należy pamiętać, że odpowiedź musi być zrozumiała i łatwa w interpretacji dla użytkownika naszego API, a z drugiej strony nie można zapomnieć o kwestiach związanych z bezpieczeństwem.
+
+Zacznijmy od przykładu.
+
+```
+GET /films/1234
+Response: 404
+```
+
+Co będzie oznaczała odpowiedź z kodem błędu 404? Może nie ma takiego zasobu, może zasób istnieje, ale nie mamy uprawnień do niego, a może ścieżka jest niewłaściwa? Dlatego część osób uważa, że należy zwracać inny kod błędu w każdym z tych przypadków.
```

Comment:
To drugie zdanie trzeba ująć bardziej literacko, bo w tym momencie brzmi trochę jakbyśmy czytali słyszane myśli.
==> Możemy podejrzewać np. brak takiego zasobu, albo np. że zasób istnieje, ale nie mamy uprawnień do niego. Inną nasuwającą się opcją jest to, że ścieżka jest niewłaściwa.